sql: add implicit SELECT FOR SHARE locking to FK checks

michae2 opened the pull request:
**explain: add transaction information to EXPLAIN ANALYZE**

Add transaction isolation, priority, and quality-of-service to the
output of `EXPLAIN ANALYZE`.

Release note (sql change): `EXPLAIN ANALYZE` output now includes:
- the isolation level of the statement's transaction
- the priority of the statement's transaction
- the quality of service level of the statement's transaction

---

**opt: do not use LockDurabilityGuaranteed under serializable isolation**

This is a follow-up from https://github.com/cockroachdb/cockroach/pull/103734.

We do not want to use guaranteed-durable (a.k.a. replicated) locking
under serializable isolation, because it prevents pipelining and other
optimizations, and is unnecessary for correctness. This commit amends
8cbc6d1360d46e3485e4828c4e9a9a85ac3d8a19 to only set durability for
`SELECT FOR UPDATE` locking under weaker isolation levels.

This means that query plans will be slightly different under different
isolation levels, and so we must add isolation level to the optimizer
memo staleness calculation.

Furthermore, this commit changes the error message added by
e633d5edb47b54a08905ae92d842625ba7805000 to be about guaranteed-durable
locking rather than `SELECT FOR UPDATE`, because in a later commit this
specific error will also be triggered by foreign key checks under weaker
isolation levels.

Informs: #100144, #100156, #100193, #100194

Release note: None

---

**opt: show locking durability in EXPLAIN (OPT) output**

Because the "guaranteed-durable locking not yet implemented" error
condition is checked in execbuilder, it prevents not only execution but
also `EXPLAIN` of queries using guaranteed-durable locking. Thankfully
`EXPLAIN (OPT)` bypasses execbuilder, and hence still works, so use this
for now to verify that we are enabling durable locking for `SELECT FOR
UPDATE` under read committed isolation.

(Note that we have not yet fixed the `SELECT FOR UPDATE` plans to use
more precise locking, that will come in a later PR.)

Informs: #100194

Release note: None

---

**sql: add implicit SELECT FOR SHARE locking to FK parent checks**

Add SELECT FOR SHARE locking to FK parent checks. Under serializable
isolation, this locking is only used when
`enable_implicit_fk_locking_for_serializable` is set. Under weaker
isolation levels (snapshot and read committed) this locking is always
used.

We only need to lock during the insertion-side FK checks, which verify
the existence of a parent row. Deletion-side FK checks verify the
non-existence of a child row, and these do not need to lock. Instead, to
prevent concurrent inserts or updates to the child that would violate
the FK constraint, we rely on the intent(s) created by the deletion
conflicting with the FK locking of those concurrent inserts or updates.

Fixes: #80683
Informs: #100156

Epic: CRDB-25322

Release note (sql change): Add a new session variable,
`enable_implicit_fk_locking_for_serializable`, which controls locking
during foreign key checks under serializable isolation. With this set to
true, foreign key checks of the referenced (parent) table, such as those
performed during an INSERT or UPDATE of the referencing (child) table,
will lock the referenced row using SELECT FOR SHARE locking. (This is
somewhat analogous to the existing `enable_implicit_select_for_update`
variable but applies to the foreign key checks of a mutation statement
instead of the initial row fetch.)

Under weaker isolation levels such as read committed, SELECT FOR SHARE
locking will always be used to ensure the database maintains the foreign
key constraint, regardless of the current setting of
`enable_implicit_fk_locking_for_serializable`.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/105857)
<!-- Reviewable:end -->

michae2 committed:
explain: add transaction information to EXPLAIN ANALYZE

Add transaction isolation, priority, and quality-of-service to the
output of `EXPLAIN ANALYZE`.

Release note (sql change): `EXPLAIN ANALYZE` output now includes:
- the isolation level of the statement's transaction
- the priority of the statement's transaction
- the quality of service level of the statement's transaction

michae2 committed:
opt: do not use LockDurabilityGuaranteed under serializable isolation

This is a follow-up from #103734.

We do not want to use guaranteed-durable (a.k.a. replicated) locking
under serializable isolation, because it prevents pipelining and other
optimizations, and is unnecessary for correctness. This commit ammends
8cbc6d1360d46e3485e4828c4e9a9a85ac3d8a19 to only set durability for
`SELECT FOR UPDATE` locking under weaker isolation levels.

This means that query plans will be slightly different under different
isolation levels, and so we must add isolation level to the optimizer
memo staleness calculation.

Furthermore, this commit changes the error message added by
e633d5edb47b54a08905ae92d842625ba7805000 to be about guaranteed-durable
locking rather than `SELECT FOR UPDATE`, because in a later commit this
specific error will also be triggered by foreign key checks under weaker
isolation levels.

Informs: #100144, #100156, #100193, #100194

Release note: None

michae2 committed:
opt: show locking durability in EXPLAIN (OPT) output

Because the "guaranteed-durable locking not yet implemented" error
condition is checked in execbuilder, it prevents not only execution but
also `EXPLAIN` of queries using guaranteed-durable locking. Thankfully
`EXPLAIN (OPT)` bypasses execbuilder, and hence still works, so use this
for now to verify that we are enabling durable locking for `SELECT FOR
UPDATE` under read committed isolation.

(Note that we have not yet fixed the `SELECT FOR UPDATE` plans to use
more precise locking, that will come in a later PR.)

Informs: #100194

Release note: None

michae2 committed:
sql: add implicit SELECT FOR SHARE locking to FK parent checks

Add SELECT FOR SHARE locking to FK parent checks. Under serializable
isolation, this locking is only used when
`enable_implicit_fk_locking_for_serializable` is set. Under weaker
isolation levels (snapshot and read committed) this locking is always
used.

We only need to lock during the insertion-side FK checks, which verify
the existence of a parent row. Deletion-side FK checks verify the
non-existence of a child row, and these do not need to lock. Instead, to
prevent concurrent inserts or updates to the child that would violate
the FK constraint, we rely on the intent(s) created by the deletion
conflicting with the FK locking of those concurrent inserts or updates.

Fixes: #80683
Informs: #100156

Epic: CRDB-25322

Release note (sql change): Add a new session variable,
`enable_implicit_fk_locking_for_serializable`, which controls locking
during foreign key checks under serializable isolation. With this set to
true, foreign key checks of the referenced (parent) table, such as those
performed during an INSERT or UPDATE of the referencing (child) table,
will lock the referenced row using SELECT FOR SHARE locking. (This is
somewhat analogous to the existing `enable_implicit_select_for_update`
variable but applies to the foreign key checks of a mutation statement
instead of the initial row fetch.)

Under weaker isolation levels such as read committed, SELECT FOR SHARE
locking will always be used to ensure the database maintains the foreign
key constraint, regardless of the current setting of
`enable_implicit_fk_locking_for_serializable`.

michae2 commented:
We've gone from "duable" to "durable" so this is now RFAL. 🙂

DrewKimball reviewed (approved):
Nice work! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> apart from a few nits.

> Reviewed 11 of 11 files at r1, 12 of 12 files at r2, 3 of 3 files at r3, 28 of 28 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-:-N_p9Hpl7diisG3L-SKN:b-y4hyk7)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @michae2, @nvanbenschoten, and @rytaft)

___
*[`pkg/sql/opt/optbuilder/mutation_builder_fk.go` line 616 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_p6Lr76xs7aE1eq7jX:-N_p6Lr76xs7aE1eq7jY:b-jd1h57) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bc66ffc22fede9349645ca351c960cf4f0d1eba/pkg/sql/opt/optbuilder/mutation_builder_fk.go#L616)):*
> ```Go
> 
> // buildOtherTableScan builds a Scan of the "other" table.
> func (h *fkCheckHelper) buildOtherTableScan(parent bool) (outScope *scope, tabMeta *opt.TableMeta) {
> ```

[nit] Maybe explain what `parent` means.
___
*[`pkg/sql/row/locking.go` line 47 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_p6QZU2EBA3uBFjrBd:-N_p6QZU2EBA3uBFjrBe:bi4t853) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bc66ffc22fede9349645ca351c960cf4f0d1eba/pkg/sql/row/locking.go#L47)):*
> ```Go
> }
> 
> // getWaitPolicy returns the configured lock wait policy to use for key-value
> ```

[nit] `getWaitPolicy` -> `GetWaitPolicy`
___
*[`pkg/sql/opt/exec/execbuilder/testdata/fk` line 417 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_p82OT2kxtHG9aGSHl:-N_p82OT2kxtHG9aGSHm:brjoryl) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bc66ffc22fede9349645ca351c960cf4f0d1eba/pkg/sql/opt/exec/execbuilder/testdata/fk#L417)):*
> ```
> 
> # Even with implicit locking enabled, we should not acquire any shared locks
> # when checking a delete from the parent.
> ```

Could you explain why the shared locks are unnecessary? Is it because inserts/updates to the child table will already check for existence of the parent row?
___
*[`pkg/sql/opt/exec/execbuilder/testdata/fk_read_committed` line 24 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_p8tpQ0piIFgIC7wsZ:-N_p8tpQ0piIFgIC7ws_:bygvn4r) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bc66ffc22fede9349645ca351c960cf4f0d1eba/pkg/sql/opt/exec/execbuilder/testdata/fk_read_committed#L24)):*
> ```
>            └── anti-join (lookup jars)
>                 ├── lookup columns are key
>                 ├── locking: for-share,duability-guaranteed
> ```

This says `duability` still :)


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
Nice work! Just a couple of questions.

> Reviewed 11 of 11 files at r1, 12 of 12 files at r2, 3 of 3 files at r3, 28 of 28 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-:-N_pK-qhEOeKoYRE9QSx:bfvcokv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @michae2, and @nvanbenschoten)

___
*[`pkg/sql/conn_executor_exec.go` line 2931 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_o0_os-feV4k3fuffT:-N_o0_os-feV4k3fuffU:b86qc48) ([raw file](https://github.com/cockroachdb/cockroach/blob/43818e898ac1d3d4fc1a231c4417ed427700ba70/pkg/sql/conn_executor_exec.go#L2931)):*
> ```Go
> 		// add qos?
> 		// add asoftime?
> 		// add readonly?
> ```

Did you mean to actually add these here? If not, this should probably be a TODO with your name / issue number
___
*[`pkg/sql/logictest/testdata/logic_test/fk_read_committed` line 35 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_pIAmp2swEHllWrvUz:-N_pIAmq4ZIhnfP_RDRZ:btqri87) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bc66ffc22fede9349645ca351c960cf4f0d1eba/pkg/sql/logictest/testdata/logic_test/fk_read_committed#L35)):*
> ```
> UPDATE cookies SET j = 2 WHERE c = 1
> 
> # Foreign key checks of the child do not require locking.
> ```

Why not? What happens if you update or delete a jar that has no cookies in it, but a concurrent RC transaction tries to insert a cookie?


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
Thanks for the reviews! Fixed the nits and added more tests and comments.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-:-N_puDUm7EjViX0XvE_H:btqq1y1)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @mgartner, @nvanbenschoten, and @rytaft)

___
*[`pkg/sql/conn_executor_exec.go` line 2931 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_o0_os-feV4k3fuffT:-N_pseteCLyDQijvbW9n:b-reptei) ([raw file](https://github.com/cockroachdb/cockroach/blob/43818e898ac1d3d4fc1a231c4417ed427700ba70/pkg/sql/conn_executor_exec.go#L2931)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Did you mean to actually add these here? If not, this should probably be a TODO with your name / issue number
</blockquote></details>

Oh, whoops, I forgot about these. Opened #107318 and added TODOs.
___
*[`pkg/sql/opt/optbuilder/mutation_builder_fk.go` line 616 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_p6Lr76xs7aE1eq7jX:-N_psv3vE4mc_TAUjn6B:b-48vilg) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bc66ffc22fede9349645ca351c960cf4f0d1eba/pkg/sql/opt/optbuilder/mutation_builder_fk.go#L616)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] Maybe explain what `parent` means.
</blockquote></details>

Good call, done.
___
*[`pkg/sql/row/locking.go` line 47 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_p6QZU2EBA3uBFjrBd:-N_psyqM0HgfWWR3DIxD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bc66ffc22fede9349645ca351c960cf4f0d1eba/pkg/sql/row/locking.go#L47)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] `getWaitPolicy` -> `GetWaitPolicy`
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/fk_read_committed` line 35 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_pIAmp2swEHllWrvUz:-N_pt-Kh1ploCH2GI18k:b9avrjh) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bc66ffc22fede9349645ca351c960cf4f0d1eba/pkg/sql/logictest/testdata/logic_test/fk_read_committed#L35)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Why not? What happens if you update or delete a jar that has no cookies in it, but a concurrent RC transaction tries to insert a cookie?
</blockquote></details>

The concurrent RC transaction will perform its *own* FK check of the parent row, which will acquire a lock on the parent jar. This lock will conflict with the update or delete of the parent jar.

This is a bit tricky so I added some comments in various places about it.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/fk` line 417 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_p82OT2kxtHG9aGSHl:-N_ptfxwDb6KCWDHhZL5:b77mkp2) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bc66ffc22fede9349645ca351c960cf4f0d1eba/pkg/sql/opt/exec/execbuilder/testdata/fk#L417)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Could you explain why the shared locks are unnecessary? Is it because inserts/updates to the child table will already check for existence of the parent row?
</blockquote></details>

Yes, exactly. Inserts/updates to the child table will do their *own* FK checks which will acquire locks on the parent row. Those locks will conflict with deletions from the parent table, preventing violations of the FK constraint.

This is a bit tricky so I added more comments about it.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/fk_read_committed` line 24 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_p8tpQ0piIFgIC7wsZ:-N_pu5p59mCAhbrXB7UA:b-alerci) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bc66ffc22fede9349645ca351c960cf4f0d1eba/pkg/sql/opt/exec/execbuilder/testdata/fk_read_committed#L24)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

This says `duability` still :)
</blockquote></details>

Gah, thank you. Done.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
Very nice! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 10 of 11 files at r1, 40 of 40 files at r5, 12 of 12 files at r6, 3 of 3 files at r7, 31 of 32 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-:-N_qNYiO-t02elNfXZ1D:bh12bql)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @mgartner, @michae2, and @rytaft)

___
*[`pkg/sql/insert_fast_path.go` line 194 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_qJm3S3qRsUG_3gRAM:-N_qJm3S3qRsUG_3gRAN:b-5nzkfx) ([raw file](https://github.com/cockroachdb/cockroach/blob/eefacc71e03b2d49947a997f582086bf7f5e5d5f/pkg/sql/insert_fast_path.go#L194)):*
> ```Go
> 		lockWaitPolicy := row.GetWaitPolicy(descpb.ToScanLockingWaitPolicy(c.Locking.WaitPolicy))
> 		// All FK check lock wait policies should be BLOCK.
> 		if r.fkBatch.Header.WaitPolicy != lockWaitPolicy {
> ```

nit: why compare against `r.fkBatch.Header.WaitPolicy` (which we assume has not been set yet) instead of directly against `lock.WaitPolicy_Block`?
___
*[`pkg/sql/vars.go` line 2805 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_qKGmI9_7TnMe1hjui:-N_qKGmI9_7TnMe1hjuj:bi6pvhm) ([raw file](https://github.com/cockroachdb/cockroach/blob/eefacc71e03b2d49947a997f582086bf7f5e5d5f/pkg/sql/vars.go#L2805)):*
> ```Go
> 
> 	// CockroachDB extension.
> 	`enable_implicit_fk_locking`: {
> ```

Should this setting mention "serializable" somehow, since setting it to false is ignored for weaker isolation levels?
___
*[`pkg/sql/opt/exec/explain/output.go` line 415 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_qG1IF2svp7sq5_NW9:-N_qG1IF2svp7sq5_NWA:b-jnqm76) ([raw file](https://github.com/cockroachdb/cockroach/blob/759944b8c2e4b29f0f8daf30e85f1def37476f55/pkg/sql/opt/exec/explain/output.go#L415)):*
> ```Go
> 	txnIsoLevel isolation.Level, txnPriority roachpb.UserPriority, txnQoSLevel sessiondatapb.QoSLevel,
> ) {
> 	ob.AddTopLevelField("isolation level", txnIsoLevel.String())
> ```

If we want to match the other fields in the output, we could use `StringLower()` here.
___
*[`pkg/sql/opt/memo/expr_format.go` line 1563 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_qHgAx-dwBQ55t7Sp0:-N_qHgAx-dwBQ55t7Sp1:byzzgiz) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a196543db5aa39849aab752cab0089a879e3b5b/pkg/sql/opt/memo/expr_format.go#L1563)):*
> ```Go
> 	case tree.LockDurabilityBestEffort:
> 	case tree.LockDurabilityGuaranteed:
> 		durability = ",durability-guaranteed"
> ```

nit: should we add `default: panic(errors.AssertionFailedf("unexpected durability"))` to catch new durability levels?
___
*[`pkg/sql/opt/optbuilder/mutation_builder_fk.go` line 650 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_qKvBCBHRwKitW4_pL:-N_qKvBCBHRwKitW4_pM:b-tu6klh) ([raw file](https://github.com/cockroachdb/cockroach/blob/eefacc71e03b2d49947a997f582086bf7f5e5d5f/pkg/sql/opt/optbuilder/mutation_builder_fk.go#L650)):*
> ```Go
> 		locking = lockingSpec{
> 			&tree.LockingItem{
> 				// TODO(michae2): Change this to ForKeyShare when it is supported.
> ```

You could set it to `row.ForKeyShare` now if you'd like. That's handled the same as `tree.ForShare` in `row.GetKeyLockingStrength`.

But also, feel free to leave this as is if you think it would be confusing to users, given that the difference doesn't matter yet.
___
*[`pkg/sql/opt/optbuilder/select.go` line 694 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_qJ58q3LTPkRbKnALN:-N_qJ58q3LTPkRbKnALO:b-v2pov8) ([raw file](https://github.com/cockroachdb/cockroach/blob/243aeb046673736730ac526186a8fb8482fddcfe/pkg/sql/opt/optbuilder/select.go#L694)):*
> ```Go
> 	if locking.isSet() {
> 		private.Locking = locking.get()
> 		if b.evalCtx.TxnIsoLevel != isolation.Serializable {
> ```

Did we want to expose a cluster setting and/or session var to allow users to opt-in to guaranteed lock durability for SFU under Serializable isolation?
___
*[`pkg/sql/sessiondatapb/local_only_session_data.proto` line 410 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_qLd1N4wIMBN42BIjV:-N_qLd1N4wIMBN42BIjW:b-femz8o) ([raw file](https://github.com/cockroachdb/cockroach/blob/eefacc71e03b2d49947a997f582086bf7f5e5d5f/pkg/sql/sessiondatapb/local_only_session_data.proto#L410)):*
> ```proto
>   // ImplicitFKLocking is true if FOR SHARE locking may be used while checking
>   // the referenced table during an insert or update to a table with a foreign
>   // key.
> ```

Same point about qualifying this somehow with the fact that it only applies to serializable.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105857)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @mgartner, @nvanbenschoten, and @rytaft)

___
*[`pkg/sql/opt/optbuilder/select.go` line 694 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_qJ58q3LTPkRbKnALN:-N_t_ry15thY2LSSG1Pc:b-3k8czk) ([raw file](https://github.com/cockroachdb/cockroach/blob/243aeb046673736730ac526186a8fb8482fddcfe/pkg/sql/opt/optbuilder/select.go#L694)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Did we want to expose a cluster setting and/or session var to allow users to opt-in to guaranteed lock durability for SFU under Serializable isolation?
</blockquote></details>

Oh, good idea! I'm thinking a session var. @nvanbenschoten do you have any preferences about the name of the setting? How about "lock_durability" or something?


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105857)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @mgartner, @michae2, and @rytaft)

___
*[`pkg/sql/opt/optbuilder/select.go` line 694 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_qJ58q3LTPkRbKnALN:-NaCq28C2oNRsRq8F_13:bpkhnbp) ([raw file](https://github.com/cockroachdb/cockroach/blob/243aeb046673736730ac526186a8fb8482fddcfe/pkg/sql/opt/optbuilder/select.go#L694)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Oh, good idea! I'm thinking a session var. @nvanbenschoten do you have any preferences about the name of the setting? How about "lock_durability" or something?
</blockquote></details>

Maybe `enable_durable_select_for_update_locks`? If possible, making the variable boolean-valued seems like the way to go, as it makes the knob easier to use.

I think we'll also want to mention something about "serializable", like we'll want to do for `enable_implicit_fk_locking` (see other comment). Maybe suffix both with `_with_serializable`, or something like that.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 40 of 40 files at r5, 12 of 12 files at r6, 3 of 3 files at r7, 32 of 32 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-:-NaDpt_48dExSNCnOcQZ:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @mgartner, and @michae2)

___
*[`pkg/sql/logictest/testdata/logic_test/fk_read_committed` line 35 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_pIAmp2swEHllWrvUz:-NaDp54w98RtxLezfjWW:b-wsvasg) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bc66ffc22fede9349645ca351c960cf4f0d1eba/pkg/sql/logictest/testdata/logic_test/fk_read_committed#L35)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

The concurrent RC transaction will perform its *own* FK check of the parent row, which will acquire a lock on the parent jar. This lock will conflict with the update or delete of the parent jar.

This is a bit tricky so I added some comments in various places about it.
</blockquote></details>

Thanks for the explanation!


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-:-NaFjFNMCHc8zmdKmf6N:bz5mair)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 3 stale) (waiting on @DrewKimball, @mgartner, @nvanbenschoten, and @rytaft)

___
*[`pkg/sql/insert_fast_path.go` line 194 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_qJm3S3qRsUG_3gRAM:-NaFgNk76Qg-3sFGP_Hl:bmn9sdb) ([raw file](https://github.com/cockroachdb/cockroach/blob/eefacc71e03b2d49947a997f582086bf7f5e5d5f/pkg/sql/insert_fast_path.go#L194)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: why compare against `r.fkBatch.Header.WaitPolicy` (which we assume has not been set yet) instead of directly against `lock.WaitPolicy_Block`?
</blockquote></details>

We're comparing against the header because there's an impedance mismatch between the `r.fkChecks[i].Locking.WaitPolicy` which is per-fk-check and the `r.fkBatch.Header.WaitPolicy` which is per-batch (one for *all* the fk checks). So just verifying that they all match whatever the batch is set to.

I changed `addFKChecks` to now explicitly set `r.fkBatch.Header.WaitPolicy` to blocking (was relying on it being implicitly set to blocking during `defer n.run.fkBatch.Reset()` in `runFKChecks`). 
___
*[`pkg/sql/vars.go` line 2805 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_qKGmI9_7TnMe1hjui:-NaFi9P2-HDqM8Ty-Ige:b-5qgjcs) ([raw file](https://github.com/cockroachdb/cockroach/blob/eefacc71e03b2d49947a997f582086bf7f5e5d5f/pkg/sql/vars.go#L2805)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should this setting mention "serializable" somehow, since setting it to false is ignored for weaker isolation levels?
</blockquote></details>

Good point, I added "_for_serializable" to be similar to Rafi's "max_retries_for_read_committed_transactions" in #107044.
___
*[`pkg/sql/opt/exec/explain/output.go` line 415 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_qG1IF2svp7sq5_NW9:-NaFiflECAZ43OuHjbj0:b-4fjqtb) ([raw file](https://github.com/cockroachdb/cockroach/blob/759944b8c2e4b29f0f8daf30e85f1def37476f55/pkg/sql/opt/exec/explain/output.go#L415)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

If we want to match the other fields in the output, we could use `StringLower()` here.
</blockquote></details>

Oh! Thank you! Done.
___
*[`pkg/sql/opt/memo/expr_format.go` line 1563 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_qHgAx-dwBQ55t7Sp0:-NaFilm61LO3Gmezk4iC:b-4c5ezp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a196543db5aa39849aab752cab0089a879e3b5b/pkg/sql/opt/memo/expr_format.go#L1563)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: should we add `default: panic(errors.AssertionFailedf("unexpected durability"))` to catch new durability levels?
</blockquote></details>

Good catch, done.
___
*[`pkg/sql/opt/optbuilder/mutation_builder_fk.go` line 650 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_qKvBCBHRwKitW4_pL:-NaFipqE4JmGfhhb_d3J:bb0ip7u) ([raw file](https://github.com/cockroachdb/cockroach/blob/eefacc71e03b2d49947a997f582086bf7f5e5d5f/pkg/sql/opt/optbuilder/mutation_builder_fk.go#L650)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

You could set it to `row.ForKeyShare` now if you'd like. That's handled the same as `tree.ForShare` in `row.GetKeyLockingStrength`.

But also, feel free to leave this as is if you think it would be confusing to users, given that the difference doesn't matter yet.
</blockquote></details>

Since this will be shown in `EXPLAIN` output, it felt wrong to show `FOR KEY SHARE` if we're not yet doing it, so I kept it this way for now.
___
*[`pkg/sql/opt/optbuilder/select.go` line 694 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_qJ58q3LTPkRbKnALN:-NaFj4yYB2DE0_CwS4aH:baxtdwq) ([raw file](https://github.com/cockroachdb/cockroach/blob/243aeb046673736730ac526186a8fb8482fddcfe/pkg/sql/opt/optbuilder/select.go#L694)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Maybe `enable_durable_select_for_update_locks`? If possible, making the variable boolean-valued seems like the way to go, as it makes the knob easier to use.

I think we'll also want to mention something about "serializable", like we'll want to do for `enable_implicit_fk_locking` (see other comment). Maybe suffix both with `_with_serializable`, or something like that.
</blockquote></details>

I started doing this but it got a little long, will do it in the next PR.
___
*[`pkg/sql/sessiondatapb/local_only_session_data.proto` line 410 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-N_qLd1N4wIMBN42BIjV:-NaFjC6ZBomg85n7xN2H:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/eefacc71e03b2d49947a997f582086bf7f5e5d5f/pkg/sql/sessiondatapb/local_only_session_data.proto#L410)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Same point about qualifying this somehow with the fact that it only applies to serializable.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Great work! I appreciate the thoughtful organization of commits - it made reviewing pleasant.

> Reviewed 10 of 11 files at r1, 40 of 40 files at r5, 12 of 12 files at r6, 3 of 3 files at r7, 32 of 32 files at r8, 52 of 52 files at r9, 12 of 12 files at r10, 3 of 3 files at r11, 32 of 32 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-:-NaIJGnW47zoTVCpVmTf:b-flqfkb)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 3 stale) (waiting on @DrewKimball and @michae2)

___
*[`pkg/sql/opt/exec/execbuilder/testdata/fk_read_committed` line 86 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-NaII0Q-46DemOaQION7:-NaII0Q-46DemOaQION8:bxy55tf) ([raw file](https://github.com/cockroachdb/cockroach/blob/eefacc71e03b2d49947a997f582086bf7f5e5d5f/pkg/sql/opt/exec/execbuilder/testdata/fk_read_committed#L86)):*
> ```
>                 └── filters (true)
> 
> # Foreign key checks of the child do not require locking.
> ```

Is this explained somewhere in more detail? Is this safe because any updates to the child rows would lock this parent row?


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
Thanks, all! This should be ready now...

bors r=DrewKimball,nvanbenschoten,rytaft,mgartner

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-:-NaJIwaTE2wMRA8__rB6:bttxbg2)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 4 stale) (waiting on @DrewKimball, @mgartner, @nvanbenschoten, and @rytaft)

___
*[`pkg/sql/opt/exec/execbuilder/testdata/fk_read_committed` line 86 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/105857#-NaII0Q-46DemOaQION7:-NaJHpxkExi3b3BfBIIC:b-7xaf0s) ([raw file](https://github.com/cockroachdb/cockroach/blob/eefacc71e03b2d49947a997f582086bf7f5e5d5f/pkg/sql/opt/exec/execbuilder/testdata/fk_read_committed#L86)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Is this explained somewhere in more detail? Is this safe because any updates to the child rows would lock this parent row?
</blockquote></details>

Yes, that's correct. It is explained in a little more detail in comments in `pkg/sql/opt/optbuilder/mutation_builder_fk.go` as well as in the 4th commit message.


<!-- Sent from Reviewable.io -->

michae2 commented:
>> Maybe enable_durable_select_for_update_locks? If possible, making the variable boolean-valued seems like the way to go, as it makes the knob easier to use.
>>
>> I think we'll also want to mention something about "serializable", like we'll want to do for enable_implicit_fk_locking (see other comment). Maybe suffix both with _with_serializable, or something like that.
>
> I started doing this but it got a little long, will do it in the next PR.

Follow-up PR is: https://github.com/cockroachdb/cockroach/pull/107749
